x/multisig/handlers.go

```diff
@@ -125,44 +125,38 @@ func (h UpdateContractMsgHandler) Deliver(ctx weave.Context, db weave.KVStore, t
 	return res, nil
 }
 
-// validate does all common pre-processing between Check and Deliver
 func (h UpdateContractMsgHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateContractMsg, error) {
-	msg, err := tx.GetMsg()
+	txmsg, err := tx.GetMsg()
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "cannot get transaction message")
 	}
 
-	updateContractMsg, ok := msg.(*UpdateContractMsg)
+	msg, ok := txmsg.(*UpdateContractMsg)
 	if !ok {
-		return nil, errors.WithType(errors.ErrInvalidMsg, msg)
+		return nil, errors.Wrapf(errors.ErrInvalidType, "%T", msg)
 	}
 
-	err = updateContractMsg.Validate()
+	// Using current version of the contract, ensure that enoguht
+	// participants with enought power/weight signed this transaction in
+	// order to run functionality that requires admin rights.
+	contract, err := h.bucket.GetContract(db, msg.ContractID)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "bucket lookup")
 	}
-
-	// load contract
-	obj, err := h.bucket.Get(db, updateContractMsg.Id)
-	if err != nil {
-		return nil, err
+	var power Weight
+	for _, p := range contract.Participants {
+		if h.auth.HasAddress(ctx, p.Signature) {
+			power += p.Power
+		}
 	}
-	if obj == nil || (obj != nil && obj.Value() == nil) {
-		return nil, errors.Wrapf(errors.ErrNotFound, contractNotFoundFmt, updateContractMsg.Id)
+	if power < contract.AdminThreshold {
+		return msg, errors.Wrapf(errors.ErrUnauthorized,
+			"%d power is not enough to administrate %q", power, msg.ContractID)
 	}
-	contract := obj.Value().(*Contract)
 
-	// retrieve sigs
-	var sigs []weave.Address
-	for _, sig := range contract.Sigs {
-		sigs = append(sigs, sig)
-	}
-
-	// check sigs
-	authenticated := x.HasNAddresses(ctx, h.auth, sigs, int(contract.AdminThreshold))
-	if !authenticated {
-		return nil, errors.Wrapf(errors.ErrUnauthorized, "contract=%X", updateContractMsg.Id)
+	if err := msg.Validate(); err != nil {
```

Comment:
I moved it after checking the weights because I want to get `ErrUnauthorized` before any validation is executed. Otherwise, you can create a transaction without signing that will be validated.
Not sure if that is a bit deal but it feels more appropriate to check permissions before running any functionality.
What do you think?

Reply:
With other handlers like cash or escrow we do input validation before auth checks. I have not checked the others but it would be nice to be consistent

Reply:
I agree. We should validate the message first.
But not modify db state until we have auth